libraries/standard/mqtt/src/mqtt.c

```diff
@@ -586,13 +789,148 @@ static MQTTStatus_t handleKeepAlive( MQTTContext_t * const pContext )
 
 /*-----------------------------------------------------------*/
 
+static bool packetStatePresent( MQTTContext_t * const pContext,
+                                uint16_t packetId,
+                                MQTTPublishState_t state )
+{
+    bool hasPublishReceived = false;
+    uint16_t packetIdFromStoredPublish = MQTT_PACKET_ID_INVALID;
+    MQTTStateCursor_t cursor = 0U;
+
+    assert( pContext != NULL );
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+
+    /* Find out if the publish packet has already received and state
+     * is maintained in the state machine. */
+    do
+    {
+        /* Query state machine for packet ids in given state. */
+        packetIdFromStoredPublish = MQTT_StateSelect( pContext,
+                                                      state,
+                                                      &cursor );
```

Comment:
We don't really *need* to use state select here. Given how `packetStatePresent` is used in the rest of the code, it seems like a function that finds if a packet ID exists regardless of state might be more useful:
```c
records = ( incomingPublish )? pContext->incomingPublishRecords : pContext->outgoingPublishRecords;
for( int i = 0; i < MQTT_STATE_ARRAY_MAX_COUNT; i++)
{
    if( records[ i ].packetId == packetId )
    {
        hasPublishReceived = true;
        *pRetrievedState = records[i].publishState;
        break;
    }
}
```